source/dofs/dof_renumbering.cc

```diff
@@ -405,26 +410,51 @@ namespace DoFRenumbering
     //
     // note that if constraints are not requested, then the 'constraints'
     // object will be empty and using it has no effect
-    IndexSet locally_relevant_dofs;
-    DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);
+    IndexSet        locally_relevant_dofs;
+    const IndexSet &locally_owned_dofs = [&]() -> const IndexSet & {
+      if (reorder_level_dofs == false)
+        {
+          DoFTools::extract_locally_relevant_dofs(dof_handler,
+                                                  locally_relevant_dofs);
+          return dof_handler.locally_owned_dofs();
+        }
+      else
+        {
+          Assert(dof_handler.n_dofs(level) != numbers::invalid_dof_index,
+                 ExcDoFHandlerNotInitialized());
+          DoFTools::extract_locally_relevant_level_dofs(dof_handler,
+                                                        level,
+                                                        locally_relevant_dofs);
+          return dof_handler.locally_owned_mg_dofs(level);
+        }
+    }();
```

Comment:
While I like the beauty of this function, I wonder whether the complication is actually necessary. Do I miss something or would the code below achieve the same thing (same number of lines, with one additional copy but so cheap that I as an HPC person would not see the need)?
```c++
    IndexSet locally_owned_dofs;
      if (reorder_level_dofs == false)
        {
          DoFTools::extract_locally_relevant_dofs(dof_handler,
                                                  locally_relevant_dofs);
          locally_owned_dofs = dof_handler.locally_owned_dofs();
        }
      else
        {
          Assert(dof_handler.n_dofs(level) != numbers::invalid_dof_index,
                 ExcDoFHandlerNotInitialized());
          DoFTools::extract_locally_relevant_level_dofs(dof_handler,
                                                        level,
                                                        locally_relevant_dofs);
          locally_owned_dofs = dof_handler.locally_owned_mg_dofs(level);
        }
```

Reply:
@kronbichler Also thank you for the feedback! I added a simple test, similar to "extract_locally_active_dofs" for the level case.
To the current implementation of the lambda function: With the current one we indeed save some memory because no complete copy of the IndexSet "locally_owned_dofs" is created, we just save a const reference to it (that's why I initially just used a pointer to const, I didn't think of creating a case dependent reference with a lambda function). 
This won't probably become the memory critical part of the program but this was indeed the main motivation behind it.
Another minor advantage is that we can store the reference as "const" and can be sure, that we don't do any changes to the IndexSet by accident.
I really can't answer the question, personally I would go the "memory saving path", but of course if the deal.II developers prefer the "simpler code" solution, I will change it to your proposal.

Reply:
I do not feel very strong about it, so I let others decide.

Regarding the memory, I see your point. However, I would not be worried about the copy because most of the time `locally_owned_dofs` occupies around 100 bytes in memory. In case a single range is used (no renumbering that cross subdomain boundaries), the by far most usual case, we have 24 byte for that range, 24 byte for the std::vector, 16 bytes for the index space size and largest range, 8 bytes for the `is_compressed` boolean (due to alignment issues), and some memory, let's say 24 bytes, for `Threads::Mutex`. Of course one could construct numberings whose memory scales with the number of locally owned dofs, by using many non-contiguous ranges, but I would not optimize for those.